File: rust/src/compiler/state.rs
use crate::value::*;

pub struct State {
    tags: Vec<String>,
    saves: Vec<String>,
}

impl State {
    pub fn new() -> State {
        State {
            tags: vec![String::new()],
            saves: vec![RESET.to_string()],
        }
    }

    pub fn push(&mut self, tag: &str, string: &str) {
        let l = self.saves.len() - 1;
        let mut s = self.saves.get(l).unwrap().clone();
        s.push_str(string);
        self.saves.push(s);
        self.tags.push(String::from(tag));
    }

    pub fn pop(&mut self) -> (Option<String>, Option<String>) {
        let a = self.tags.pop();
        let b = self.saves.pop();
        (a, b)
    }

    pub fn current_tag(&self) -> &str {
        let l = self.tags.len() - 1;
        self.tags.get(l).unwrap()
    }

    pub fn current_save(&self) -> &str {
        let l = self.saves.len() - 1;
        self.saves.get(l).unwrap()
    }
}

File: javascript/src/compiler/state.ts
import { RESET } from "../value"

export class State {
    tags: string[]
    saves: string[]

    constructor() {
        this.tags = [""]
        this.saves = [RESET]
    }

    push(tag: string, string: string) {
        const l = this.saves.length - 1
        const s = this.saves[l]
        this.saves.push(s + string)
        this.tags.push(tag)
    }

    pop(): string[] {
        const a = this.tags.pop()
        const b = this.saves.pop()
        return [a, b]
    }

    currentTag(): string {
        const l = this.tags.length - 1
        return this.tags[l]
    }

    currentSave(): string {
        const l = this.saves.length - 1
        return this.saves[l]
    }
}